File: examples/university-demo/src/main/java/io/axoniq/demo/university/faculty/write/StudentId.java
package io.axoniq.demo.university.faculty.write;

import java.util.UUID;

public record StudentId(String raw) {

    private final static String ENTITY_TYPE = "Student";

    public StudentId {
        if (raw == null || raw.isBlank()) {
            throw new IllegalArgumentException("Course ID cannot be null or empty");
        }
        raw = withType(raw);
    }

    public static StudentId of(String raw) {
        return new StudentId(raw);
    }

    public static StudentId random() {
        return new StudentId(UUID.randomUUID().toString());
    }

    @Override
    public String toString() {
        return raw;
    }

    private static String withType(String id) {
        return id.startsWith(ENTITY_TYPE + ":") ? id : ENTITY_TYPE + ":" + id;
    }
}

File: examples/university-demo-kotlin/src/main/kotlin/shared/ids/StudentId.kt
@file:OptIn(ExperimentalStdlibApi::class)

package io.axoniq.demo.university.shared.ids

import java.util.*

@JvmExposeBoxed
@JvmInline
value class StudentId(val value: String) {
  companion object {
    fun random() = StudentId(UUID.randomUUID().toString())
  }

  override fun toString() = value
}

